test/src/unit-capi-dynamic.cc

```diff
@@ -0,0 +1,841 @@
+        rc = tiledb_query_set_buffer_var(
+            ctx_,
+            query,
+            test_query_buffer.name_.c_str(),
+            static_cast<uint64_t*>(test_query_buffer.buffer_),
+            test_query_buffer.buffer_size_,
+            test_query_buffer.buffer_var_,
+            test_query_buffer.buffer_var_size_);
+        REQUIRE(rc == TILEDB_OK);
+      }
+    } else {
+      if (test_query_buffer.buffer_var_ == nullptr) {
+        rc = tiledb_query_set_buffer_nullable(
+            ctx_,
+            query,
+            test_query_buffer.name_.c_str(),
+            test_query_buffer.buffer_,
+            test_query_buffer.buffer_size_,
+            test_query_buffer.buffer_validity_,
+            test_query_buffer.buffer_validity_size_);
+        REQUIRE(rc == TILEDB_OK);
+      } else {
+        rc = tiledb_query_set_buffer_var_nullable(
+            ctx_,
+            query,
+            test_query_buffer.name_.c_str(),
+            static_cast<uint64_t*>(test_query_buffer.buffer_),
+            test_query_buffer.buffer_size_,
+            test_query_buffer.buffer_var_,
+            test_query_buffer.buffer_var_size_,
+            test_query_buffer.buffer_validity_,
+            test_query_buffer.buffer_validity_size_);
+        REQUIRE(rc == TILEDB_OK);
+      }
+    }
+  }
+
+  // Set the subarray to read.
+  rc = tiledb_query_set_subarray(ctx_, query, subarray);
+  REQUIRE(rc == TILEDB_OK);
+
+  // Submit the query.
+  rc = tiledb_query_submit(ctx_, query);
+  REQUIRE(rc == TILEDB_OK);
+
+  // Finalize the query, a no-op for non-global writes.
+  rc = tiledb_query_finalize(ctx_, query);
+  REQUIRE(rc == TILEDB_OK);
+
+  // Clean up
+  rc = tiledb_array_close(ctx_, array);
+  REQUIRE(rc == TILEDB_OK);
+  tiledb_array_free(&array);
+  tiledb_query_free(&query);
+}
+
+void DynamicArrayFx::test_dynamic_array(
+    const test_attr_t test_attr,
+    const vector<test_dim_t>& test_dims,
+    tiledb_array_type_t array_type,
+    tiledb_layout_t cell_order,
+    tiledb_layout_t tile_order,
+    tiledb_layout_t write_order,
+    tiledb_encryption_type_t encryption_type) {
+  const string array_name = "dynamic_array";
+
+  // Skip row-major and col-major writes for sparse arrays.
+  if (array_type == TILEDB_SPARSE &&
+      (write_order == TILEDB_ROW_MAJOR || write_order == TILEDB_COL_MAJOR)) {
+    return;
+  }
+
+  // String_ascii, float32, and float64 types can only be
+  // written to sparse arrays.
+  if (array_type == TILEDB_DENSE && (test_attr.type_ == TILEDB_STRING_ASCII ||
+                                     test_attr.type_ == TILEDB_FLOAT32 ||
+                                     test_attr.type_ == TILEDB_FLOAT64)) {
+    return;
+  }
+
+  for (auto& test_dim : test_dims) {
+    // Create the array.
+    create_array(
+        array_name,
+        array_type,
+        test_dim,
+        test_attr,
+        cell_order,
+        tile_order,
+        encryption_type);
+  }
+
+  // Calculate the write buffer sizes.
```

Comment:
Let's clarify that we're calculating dimension ranges, we shouldn't use the word "size" here because we don't incorporate the dimension type size. How about `vector<uint64_t> dimension_ranges;`?